core/model/__init__.py

```diff
@@ -334,38 +334,40 @@ def initialize(cls, url, initialize_data=True, initialize_schema=True):
             return engine, engine.connect()
 
         engine = cls.engine(url)
-        if initialize_schema:
-            cls.initialize_schema(engine)
-        connection = engine.connect()
-
-        # Check if the recursive equivalents function exists already.
-        query = (
-            select([literal_column("proname")])
-            .select_from(table("pg_proc"))
-            .where(literal_column("proname") == "fn_recursive_equivalents")
-        )
-        result = connection.execute(query)
-        result = list(result)
-
-        # If it doesn't, create it.
-        if not result and initialize_data:
-            resource_file = os.path.join(
-                cls.resource_directory(), cls.RECURSIVE_EQUIVALENTS_FUNCTION
+        with engine.connect() as connection:
+            tx = connection.begin()
+            connection.execute("LOCK TABLE pg_proc IN ACCESS EXCLUSIVE MODE;")
+
+            if initialize_schema:
+                cls.initialize_schema(connection)
+
+            # Check if the recursive equivalents function exists already.
+            query = (
+                select([literal_column("proname")])
+                .select_from(table("pg_proc"))
+                .where(literal_column("proname") == "fn_recursive_equivalents")
             )
-            if not os.path.exists(resource_file):
-                raise OSError(
-                    "Could not load recursive equivalents function from %s: file does not exist."
-                    % resource_file
+            result = connection.execute(query)
+            result = list(result)
+
+            # If it doesn't, create it.
+            if not result and initialize_data:
+                resource_file = os.path.join(
+                    cls.resource_directory(), cls.RECURSIVE_EQUIVALENTS_FUNCTION
                 )
-            sql = open(resource_file).read()
-            connection.execute(sql)
+                if not os.path.exists(resource_file):
+                    raise OSError(
+                        "Could not load recursive equivalents function from %s: file does not exist."
+                        % resource_file
+                    )
+                sql = open(resource_file).read()
+                connection.execute(sql)
 
-        if initialize_data:
-            session = Session(connection)
-            cls.initialize_data(session)
+            if initialize_data:
+                session = Session(connection)
+                cls.initialize_data(session)
 
-        if connection:
-            connection.close()
+            tx.commit()
```

Comment:
I'm not 100% this line is necessary.  Does closing a connection automatically commit any associated transactions?

Reply:
If `autocommit` is on, it does. I think it's best to explicitly commit (or rollback), where appropriate.